### DISCUSSION TITLE
Guided Exercise - React App with API Data Fetching and Routing - Amazezon

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#174] [Tasks]
Let's create a React app called "Amazezon" (a playful spin on Amazon) that fetches product data from the ReqRes.in API. Users can view a list of products, add them to their cart, and increment or decrement the quantity of each item. The app will also implement React Router for navigation between different pages. This exercise will involve fetching data using Axios, managing state variables, implementing useEffect with and without a watchlist, creating reusable components, and styling with custom CSS. Here are the detailed step-by-step instructions:

Step 1: Setup
- Create a new directory for your project.
- Initialize a new React app using Create React App by running the command: `npx create-react-app amazezon`.
- Navigate into the project directory: `cd amazezon`.
- Start the development server: `npm start`.
- Open your browser and go to `http://localhost:3000` to see the default React app running.

Step 2: Install dependencies
- Open your terminal and navigate to the project directory if you're not already there.
- Install Axios, React Router, and Bootstrap by running the command: `npm install axios react-router-dom bootstrap`.

Step 3: Clear the default project
- Open the `src/App.js` file and remove the existing code.
- Replace it with the following code:
```jsx
import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { BrowserRouter as Router, Switch, Route, Link } from 'react-router-dom';
import Home from './pages/Home';
import ProductDetail from './pages/ProductDetail';
import Cart from './pages/Cart';
import './App.css';

function App() {
  const [products, setProducts] = useState([]);
  const [cart, setCart] = useState([]);

  useEffect(() => {
    fetchProducts();
  }, []);

  const fetchProducts = async () => {
    try {
      const response = await axios.get('https://reqres.in/api/products');
      setProducts(response.data.data);
    } catch (error) {
      console.log('Error fetching products:', error);
    }
  };

  const addToCart = (product) => {
    const existingProduct = cart.find((item) => item.id === product.id);
    if (existingProduct) {
      setCart((prevCart) =>
        prevCart.map((item) =>
          item.id === product.id ? { ...item, quantity: item.quantity + 1 } : item
        )
      );
    } else {
      setCart((prevCart) => [...prevCart, { ...product, quantity: 1 }]);
    }
  };

  const incrementQuantity = (productId) => {
    setCart((prevCart) =>
      prevCart.map((item) =>
        item.id === productId ? { ...item, quantity: item.quantity + 1 } : item
      )
    );
  };

  const decrementQuantity = (productId) => {
    setCart((prevCart) =>
      prevCart.map((item) =>
        item.id === productId ? { ...item, quantity: Math.max(item.quantity - 1, 0) } : item
      )
    );
  };

  const removeItem = (productId) => {
    setCart((prevCart) => prevCart.filter((item) => item.id !== productId));
  };

  return (
    <Router>
      <div className="container">
        <nav className="navbar navbar-expand-lg navbar-light bg-light">
          <Link to="/" className="navbar-brand">
            Amazezon
          </Link>
          <ul className="navbar-nav">
            <li className="nav-item">
              <Link to="/cart" className="nav-link">
                Cart ({cart.length})
              </Link>
            </li>
          </ul>
        </nav>
        <Switch>
          <Route exact path="/">
            <Home products={products} addToCart={addToCart} />
          </Route>
          <Route path="/product/:id">
            <ProductDetail products={products} addToCart={addToCart} />
          </Route>
          <Route path="/cart">
            <Cart
              cart={cart}
              incrementQuantity={incrementQuantity}
              decrementQuantity={decrementQuantity}
              removeItem={removeItem}
            />
          </Route>
        </Switch>
      </div>
    </Router>
  );
}

export default App;
```

Step 4: Create Home page
- Create a new directory called `pages` inside the `src` directory.
- Inside the `pages` directory, create a new file called `Home.js`.
- Open `Home.js` and add the following code:
```jsx
import React from 'react';
import { Link } from 'react-router-dom';

function Home({ products, addToCart }) {
  return (
    <div>
      <h1>Welcome to Amazezon</h1>
      <div className="row">
        {products.map((product) => (
          <div className="col-md-4 mb-4" key={product.id}>
            <div className="card">
              <img src={product.avatar} className="card-img-top" alt={product.name} />
              <div className="card-body">
                <h5 className="card-title">{product.name}</h5>
                <p className="card-text">{product.description}</p>
                <button
                  className="btn btn-primary"
                  onClick={() => addToCart(product)}
                >
                  Add to Cart
                </button>
                <Link to={`/product/${product.id}`} className="btn btn-link">
                  View Details
                </Link>
              </div>
            </div>
          </div>
        ))}
      </div>
    </div>
  );
}

export default Home;
```

Step 5: Create ProductDetail page
- Inside the `pages` directory, create a new file called `ProductDetail.js`.
- Open `ProductDetail.js` and add the following code:
```jsx
import React from 'react';
import { useParams, Link } from 'react-router-dom';

function ProductDetail({ products, addToCart }) {
  const { id } = useParams();
  const product = products.find((product) => product.id === parseInt(id));

  if (!product) {
    return <h3>Product not found</h3>;
  }

  return (
    <div>
      <div className="card">
        <img src={product.avatar} className="card-img-top" alt={product.name} />
        <div className="card-body">
          <h5 className="card-title">{product.name}</h5>
          <p className="card-text">{product.description}</p>
          <button className="btn btn-primary" onClick={() => addToCart(product)}>
            Add to Cart
          </button>
          <Link to="/" className="btn btn-link">
            Back to Home
          </Link>
        </div>
      </div>
    </div>
  );
}

export default ProductDetail;
```

Step 6: Create Cart page
- Inside the `pages` directory, create a new file called `Cart.js`.
- Open `Cart.js` and add the following code:
```jsx
import React from 'react';
import { Link } from 'react-router-dom';

function Cart({ cart, incrementQuantity, decrementQuantity, removeItem }) {
  return (
    <div>
      <h1>Cart</h1>
      {cart.length === 0

? (
        <p>Your cart is empty.</p>
      ) : (
        <>
          <table className="table">
            <thead>
              <tr>
                <th>Product</th>
                <th>Quantity</th>
                <th>Actions</th>
              </tr>
            </thead>
            <tbody>
              {cart.map((item) => (
                <tr key={item.id}>
                  <td>{item.name}</td>
                  <td>
                    <button
                      className="btn btn-sm btn-primary me-2"
                      onClick={() => decrementQuantity(item.id)}
                    >
                      -
                    </button>
                    {item.quantity}
                    <button
                      className="btn btn-sm btn-primary ms-2"
                      onClick={() => incrementQuantity(item.id)}
                    >
                      +
                    </button>
                  </td>
                  <td>
                    <button
                      className="btn btn-sm btn-danger"
                      onClick={() => removeItem(item.id)}
                    >
                      Remove
                    </button>
                  </td>
                </tr>
              ))}
            </tbody>
          </table>
          <div className="text-end">
            <Link to="/" className="btn btn-link">
              Continue Shopping
            </Link>
            <button className="btn btn-primary">Checkout</button>
          </div>
        </>
      )}
    </div>
  );
}

export default Cart;
```

Step 7: Create custom CSS
- Create a new file called `App.css` in the `src` directory.
- Open `App.css` and add the following code:
```css
.container {
  margin-top: 2rem;
}

.navbar-brand {
  font-weight: bold;
}

.nav-link {
  color: #333;
}

.card {
  margin-bottom: 1rem;
}

.btn-primary {
  margin-right: 0.5rem;
}

.btn-link {
  color: #007bff;
}

.table {
  margin-top: 1rem;
}
```

Step 8: Run the app
- Make sure the development server is running (`npm start`).
- Open your browser and go to `http://localhost:3000`.
- You should see the "Amazezon" heading, a navigation bar, and a list of products on the home page.
- Clicking on "Add to Cart" will add the product to the cart.
- The cart page will display the added items with quantity and options to increment, decrement, or remove them.
- The "Checkout" button is not functional in this exercise.
- Clicking on "View Details" will navigate to the product detail page.
- The product detail page will display the name, description, and an option to add the product to the cart.
- The "Back to Home" link will navigate back to the home page.

Congratulations! You have created the "Amazezon" React app. Users can browse products, add them to the cart, and manage the quantity of items in the cart. The app utilizes useState and useEffect for managing state and fetching data. React Router enables navigation between different pages. Bootstrap and custom CSS provide a visually appealing and responsive user interface. Feel free to further customize the styles, add additional features, or incorporate more interesting interactions to make it even more engaging and fun. Enjoy exploring and shopping with your magical Amazezon app!

**Comment by chintan12** [comment]
<img width="960" alt="Screenshot 2023-07-09 210852" src="https://github.com/akash-coded/mern/assets/11799378/05d51549-ccc5-4a9b-95ce-3b0584fcc5c7">
<img width="958" alt="Screenshot 2023-07-09 210902" src="https://github.com/akash-coded/mern/assets/11799378/22d1fe2f-a3b6-4e2b-ba91-57221d911212">
<img width="956" alt="Screenshot 2023-07-09 210949" src="https://github.com/akash-coded/mern/assets/11799378/9e053774-ddc3-4cfe-b11a-8a4e54acc67e">

**Comment by Praveen02062000** [comment]
![Screenshot (7)](https://github.com/akash-coded/mern/assets/137034312/7b567c04-560e-412b-95b6-a9162fff7758)
![Screenshot (8)](https://github.com/akash-coded/mern/assets/137034312/a02e43d0-9c0d-4a44-b155-08410b7f0cac)
![Screenshot (5)](https://github.com/akash-coded/mern/assets/137034312/c98d4c5a-ac39-4f77-b14a-d52f99a8ad33)
![Screenshot (6)](https://github.com/akash-coded/mern/assets/137034312/90e1b4c2-74fc-46eb-bbe3-9a1f168f3dc4)

**Reply by Praveen02062000** [reply]
<Routes>
          <Route exact path="/" element={  <Home products={products} addToCart={addToCart} />}/>
          
          {/* </Route> */}
          <Route path="/product/:id" element={<ProductDetail products={products} addToCart={addToCart} />}/>
            
     
          <Route path="/cart" element={ <Cart
              cart={cart}
              incrementQuantity={incrementQuantity}
              decrementQuantity={decrementQuantity}
              removeItem={removeItem}
            />}/>
           
          
        </Routes>

i used Routes.

**Comment by rishab1616** [comment]
![image](https://github.com/akash-coded/mern/assets/52340386/1cb65511-9abb-454b-9d95-c0b764ab746a)

**Reply by biswa348** [reply]
how you did it mine blank page is coming

**Reply by rishab1616** [reply]
@biswa348  Home page?

**Reply by biswa348** [reply]
yes

**Reply by rishab1616** [reply]
Just imported the code as given

**Comment by SagarTChandrakanth** [comment]
![Screenshot from 2023-07-09 21-09-54](https://github.com/akash-coded/mern/assets/110445544/95e99c19-95e7-4c1c-acdb-39c88581b76d)

**Comment by Sheetal-dh** [comment]
![image](https://github.com/akash-coded/mern/assets/91683804/09a88420-e72e-4c08-aa2c-cfd575e35e0a)

**Reply by Sheetal-dh** [reply]
![image](https://github.com/akash-coded/mern/assets/91683804/3f1f653f-4b9a-4b1c-867d-389eb47a8427)

**Comment by Rutud2120** [comment]
![image](https://github.com/akash-coded/mern/assets/104456697/33751b8c-ef9a-4a84-8334-0bfb39eec5e4)

**Comment by mightyrsm** [comment]
<img width="722" alt="Screenshot 2023-07-09 at 9 16 33 PM" src="https://github.com/akash-coded/mern/assets/136921146/b44606a1-73c8-477e-bfd0-e321062aa9c9">

**Comment by Sanket866** [comment]
<img width="196" alt="prasad" src="https://github.com/akash-coded/mern/assets/120538359/8f91c61d-b0db-447a-a379-3581253d0741">
![Uploading prasad.png…]()

**Comment by Karthikkevin** [comment]
![Screenshot (161)](https://github.com/akash-coded/mern/assets/33948674/26db0f14-955e-4444-874d-044984b49d83)
![Screenshot (162)](https://github.com/akash-coded/mern/assets/33948674/c7902cb4-1581-4013-9cf5-2ed9312e9c33)
![Screenshot (163)](https://github.com/akash-coded/mern/assets/33948674/32dfa046-039e-4bb6-a8f7-c4fc137fe24d)

**Comment by pooja744** [comment]
![image](https://github.com/akash-coded/mern/assets/138996416/67f857ce-246a-47c4-a88e-71119c58ddef)

**Comment by anoop-manukumar** [comment]
![tmp_75f2fb47-4e68-4286-9450-c35ed20f7b18](https://github.com/akash-coded/mern/assets/138996605/65ddefd1-db15-47da-86f4-c57ef0ecbb0e)

**Comment by madhuriS3008** [comment]
![image](https://github.com/akash-coded/mern/assets/85974647/59d2b1b8-3252-4bfc-9c01-f9cf1b63a148)
![image](https://github.com/akash-coded/mern/assets/85974647/bc1e980b-efd0-491d-a345-85c140787a58)
![image](https://github.com/akash-coded/mern/assets/85974647/7c0ca1ff-eb1c-4ff4-80ba-5e849ebcb474)

As I installed latest react-router, made few changes to Routes, Route and removed Switch.
To apply Bootstrap, I added "import "bootstrap/dist/css/bootstrap.min.css";" in index.js
Cart total changed to display total quantity of items present in cart using reduce method.

**Reply by madhuriS3008** [reply]
1. As I installed latest react-router, made few changes to Routes, Route and removed Switch.
2. To apply Bootstrap, I added --- import "bootstrap/dist/css/bootstrap.min.css"; -- in index.js
3. Cart total changed to display total quantity of items present in cart using reduce method.

**Reply by Rutud2120** [reply]
Could you please upload the code changes for point 1? Was getting issues at my end.

**Reply by madhuriS3008** [reply]
in App.js

![image](https://github.com/akash-coded/mern/assets/85974647/d4a03685-9bb8-4e7b-bc38-e26e6e15d6eb)

**Reply by nandini-naik09** [reply]
Can you please explain point 3 changes? or upload the code change?

**Comment by kkpai** [comment]
![Amazezon](https://github.com/akash-coded/mern/assets/15033880/f1697a6c-4c7b-4a69-92b3-c6ea89bda616)

**Comment by parameshkomati** [comment]
![image](https://github.com/akash-coded/mern/assets/47549898/04462b00-f3d2-4c25-8b06-7249f7efe6cc)

**Comment by anantsinha7798** [comment]
![IMG-20230709-WA0056](https://github.com/akash-coded/mern/assets/52276093/7f0c4e18-bf45-4d05-8d9e-83e2b9a8d273)

**Reply by KailashGareja** [reply]
import 'bootstrap/dist/css/bootstrap.min.css' this css in index.js first then bootstrap will work

**Comment by KailashGareja** [comment]
import 'bootstrap/dist/css/bootstrap.min.css' this css in index.js first then bootstrap will work

**Comment by KailashGareja** [comment]
![image](https://github.com/akash-coded/mern/assets/136106221/4353adbf-8dd0-485f-905b-9539ebc66bfb)

**Comment by saranyakarthi** [comment]
![homepage](https://github.com/akash-coded/mern/assets/6057180/b7071149-c4ab-4330-b158-d232deaf1787)

**Comment by Sagar1006-dev** [comment]
![2023-07-09 (1)](https://github.com/akash-coded/mern/assets/129708548/6427de80-358b-4c9d-bcfc-e377a0a349a7)
![2023-07-10 (1)](https://github.com/akash-coded/mern/assets/129708548/42d4835a-fb7c-4cff-a3d4-fc655aee7ef2)
![Uploading 2023-07-10.png…]()

**Comment by KavyaKetha** [comment]
![image](https://github.com/akash-coded/mern/assets/12147689/611db11b-b86f-49b7-ba0c-477c7b8ef526)
![image](https://github.com/akash-coded/mern/assets/12147689/a9834d44-966f-4bf2-8480-385d948cabe3)
![image](https://github.com/akash-coded/mern/assets/12147689/be1099c9-2890-4a96-9483-965f72d98d23)

**Comment by pawarprathmesh333** [comment]
![image](https://github.com/akash-coded/mern/assets/71115267/74553da0-b07b-4b60-8a93-5bb2031885a9)
![image](https://github.com/akash-coded/mern/assets/71115267/04cdecd7-863a-4a27-8f08-2bd4225e5f28)

**Comment by PragavathiVadivel** [comment]
![image](https://github.com/akash-coded/mern/assets/137114388/9e5f71ea-8953-4daf-bc8f-63a20bb3de98)

**Comment by PragavathiVadivel** [comment]
![image](https://github.com/akash-coded/mern/assets/137114388/4f4c1ebe-df80-40c5-8bfb-7ca6131f24c2)

**Comment by PragavathiVadivel** [comment]
Hi Akash, As I am using react-router-dom 6. I am getting problem in switch. so just I downsized to 5. It specifies the error as
" could not find the module for react-router dom ". then I saved @types.

No errors but the Link is not working. Then changed the switch and everything . The above output is implemented using Routes,Route-which is the part of react-router-dom 6.

Got the above output. but the cart is resetting to 0 when  navigating to different components. trying to solve

**Comment by vk0027** [comment]
![image](https://github.com/akash-coded/mern/assets/123943655/226f69fa-b6ee-47a5-bc08-3da54b2384c9)

**Comment by starvikas1988** [comment]
![image](https://github.com/akash-coded/mern/assets/70792693/eecfe641-2fcc-4421-882e-bb56884f7ca6)
![image](https://github.com/akash-coded/mern/assets/70792693/ff6a59e3-db57-44ec-98f5-c7d5c34450e6)

**Reply by starvikas1988** [reply]
# App.js
```js
import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { BrowserRouter as Router, Routes, Route, Link } from 'react-router-dom';
import Home from './pages/Home';
import ProductDetail from './pages/ProductDetail';
import Cart from './pages/Cart';
import './App.css';
import './bootstrap.min.css';

function App() {
  const [products, setProducts] = useState([]);
  const [cart, setCart] = useState([]);

  useEffect(() => {
    fetchProducts();
  }, []);

  const fetchProducts = async () => {
    try {
      const response = await axios.get('https://reqres.in/api/products');
      setProducts(response.data.data);
    } catch (error) {
      console.log('Error fetching products:', error);
    }
  };

  const addToCart = (product) => {
    const existingProduct = cart.find((item) => item.id === product.id);
    if (existingProduct) {
      setCart((prevCart) =>
        prevCart.map((item) =>
          item.id === product.id ? { ...item, quantity: item.quantity + 1 } : item
        )
      );
    } else {
      setCart((prevCart) => [...prevCart, { ...product, quantity: 1 }]);
    }
  };

  const incrementQuantity = (productId) => {
    setCart((prevCart) =>
      prevCart.map((item) =>
        item.id === productId ? { ...item, quantity: item.quantity + 1 } : item
      )
    );
  };

  const decrementQuantity = (productId) => {
    setCart((prevCart) =>
      prevCart.map((item) =>
        item.id === productId ? { ...item, quantity: Math.max(item.quantity - 1, 0) } : item
      )
    );
  };

  const removeItem = (productId) => {
    setCart((prevCart) => prevCart.filter((item) => item.id !== productId));
  };

  return (
    <Router>
      <div className="container">
        <nav className="navbar navbar-expand-lg navbar-light bg-light">
          <Link to="/" className="navbar-brand">
            Amazezon
          </Link>
          <ul className="navbar-nav">
            <li className="nav-item">
              <Link to="/cart" className="nav-link">
                Cart ({cart.length})
              </Link>
            </li>
          </ul>
        </nav>
        <Routes>
        <Route exact path="/" element={<Home products={products} addToCart={addToCart} />} />
         
          <Route path="/product/:id" element={<ProductDetail products={products} addToCart={addToCart} />}/>
            
          <Route path="/cart" element={<Cart
              cart={cart}
              incrementQuantity={incrementQuantity}
              decrementQuantity={decrementQuantity}
              removeItem={removeItem}
            />}/>
            
        </Routes>
      </div>
    </Router>
  );
}

export default App;
```

**Reply by starvikas1988** [reply]
![image](https://github.com/akash-coded/mern/assets/70792693/dc94b5ac-933e-48a7-ad6c-499b63953926)

**Comment by RevathiSundaramoorthy** [comment]
![image](https://github.com/akash-coded/mern/assets/126975777/4004f96e-d036-4f11-9eeb-12f8e93d948e)
![image](https://github.com/akash-coded/mern/assets/126975777/c666aa3c-ce61-46d4-98ba-82d965d6f147)
![image](https://github.com/akash-coded/mern/assets/126975777/24a35d9a-c845-4b67-ba0d-5b806e986ce8)

**Comment by suhasini130511** [comment]
import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { BrowserRouter as Router, Switch, Route, Link } from 'react-router-dom';
import Home from './pages/Home';
import ProductDetail from './pages/ProductDetail';
import Cart from './pages/Cart';
import './App.css';

function App() {
  const [products, setProducts] = useState([]);
  const [cart, setCart] = useState([]);

  useEffect(() => {
    fetchProducts();
  }, []);

  const fetchProducts = async () => {
    try {
      const response = await axios.get('https://reqres.in/api/products');
      setProducts(response.data.data);
    } catch (error) {
      console.log('Error fetching products:', error);
    }
  };

  const addToCart = (product) => {
    const existingProduct = cart.find((item) => item.id === product.id);
    if (existingProduct) {
      setCart((prevCart) =>
        prevCart.map((item) =>
          item.id === product.id ? { ...item, quantity: item.quantity + 1 } : item
        )
      );
    } else {
      setCart((prevCart) => [...prevCart, { ...product, quantity: 1 }]);
    }
  };

  const incrementQuantity = (productId) => {
    setCart((prevCart) =>
      prevCart.map((item) =>
        item.id === productId ? { ...item, quantity: item.quantity + 1 } : item
      )
    );
  };

  const decrementQuantity = (productId) => {
    setCart((prevCart) =>
      prevCart.map((item) =>
        item.id === productId ? { ...item, quantity: Math.max(item.quantity - 1, 0) } : item
      )
    );
  };

  const removeItem = (productId) => {
    setCart((prevCart) => prevCart.filter((item) => item.id !== productId));
  };

  return (
    <Router>
      <div className="container">
        <nav className="navbar navbar-expand-lg navbar-light bg-light">
          <Link to="/" className="navbar-brand">
            Amazezon
          </Link>
          <ul className="navbar-nav">
            <li className="nav-item">
              <Link to="/cart" className="nav-link">
                Cart ({cart.length})
              </Link>
            </li>
          </ul>
        </nav>
        <Switch>
          <Route exact path="/">
            <Home products={products} addToCart={addToCart} />
          </Route>
          <Route path="/product/:id">
            <ProductDetail products={products} addToCart={addToCart} />
          </Route>
          <Route path="/cart">
            <Cart
              cart={cart}
              incrementQuantity={incrementQuantity}
              decrementQuantity={decrementQuantity}
              removeItem={removeItem}
            />
          </Route>
        </Switch>
      </div>
    </Router>
  );
}

export default App;
